### PR TITLE
Cooler mobs

### PR DESCRIPTION
**Pull request by funkychicken493**
Working on a very cool mob randomizer system thingie. Feel free to critique as I go.

### CODE REVIEW

**Comment by funkychicken493** [comment]
I forgot to mention, this is based HEAVILY on [Adriftus'](https://github.com/Adriftus-Studios/network-script-data) system, so full credit goes to them, I'm just rewriting it for practice.

File: plugins/Denizen/scripts/mobs/modifiers/assignment.dsc [hunk 1]
```diff
@@ -5,11 +5,17 @@ mob_modifiers_spawned_entity:
...
+            - define selected_suffixes <[valid_suffixes].random[<[build].get[suffixes]>]>
+            - foreach <[selected_prefixes]> as:prefix:
                 - flag <context.entity> mob_modifiers.<[prefix]>
-            - foreach <[valid_suffixes].random[<[build].get[suffixes]>]> as:suffix:
-                - flag <context.entity> mob_modifiers.<[suffix]>
\ No newline at end of file
+            - foreach <[selected_suffixes]> as:suffix:
+                - flag <context.entity> mob_modifiers.<[suffix]>
+            - define color <&color[<[build].get[color]>]>
+            - define name <[color]><[selected_prefixes].parse_tag[<[parse_value].proc[snake_case_to_title_case]>].space_separated> <&f><context.entity.custom_name.if_null[<context.entity.entity_type.proc[snake_case_to_title_case]>]> <[color]><[selected_suffixes].parse_tag[<[parse_value].proc[snake_case_to_title_case]>].space_separated>
```
Comment (acikek) [inline plugins/Denizen/scripts/mobs/modifiers/assignment.dsc]: You could split up this mega-tag

File: plugins/Denizen/scripts/mobs/modifiers/suffixes/exploding.dsc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+mob_mod_exploding:
```
Comment (acikek) [inline plugins/Denizen/scripts/mobs/modifiers/suffixes/exploding.dsc]: I feel like these modifiers are short enough to have `suffixes.dsc`/`prefixes.dsc` files, then make new files for bigger ones

**Comment by acikek** [comment]
Really nice code so far :+1:

File: plugins/Denizen/scripts/mobs/modifiers/suffixes/exploding.dsc (hunk 2 above)
Comment (funkychicken493) [reply plugins/Denizen/scripts/mobs/modifiers/suffixes/exploding.dsc]: I was going off of [Adriftus'](https://github.com/Adriftus-Studios/network-script-data) system with this, but yeah you're right, will do.

- [x] Consolidate small modifiers

File: plugins/Denizen/scripts/mobs/modifiers/assignment.dsc (hunk 1 above)
Comment (funkychicken493) [reply plugins/Denizen/scripts/mobs/modifiers/assignment.dsc]: Yeah, I'll split it into `formatted_prefix` and then `formatted_suffix` that way the line will be shorter and more readable.

- [x] Carve mega-tag

**Comment by funkychicken493** [comment]
Pending testing on my end.

- [x] Test random spawning
- [x] Check that mobs spawn with names
- [x] Ensure test modifiers work
- [x] Ensure rare and common builds both work (will require more prefix modifiers to be created)

File: plugins/Denizen/scripts/utility/integerize.dsc [hunk 3]
```diff
@@ -0,0 +1,31 @@
+integerize:
+    type: procedure
+    debug: false
+    description: Multiplies a decimal or integer by 10 until it is an integer.
```
Comment (funkychicken493) [inline plugins/Denizen/scripts/utility/integerize.dsc:4]: I know this description looks dumb, but this is literally what it does. If you input an integer it returns the inputted integer.

File: plugins/Denizen/scripts/mobs/general_effects/shelled.dsc [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Shelled enemies will absorb a number of hits (specifically hits, not damage) and then break.
+
+shelled_events:
+    type: world
+    debug: false
+    events:
+        on entity_flagged:shelled takes damage:
+            - if <context.entity.flag[shelled].is_integer.not>:
```
Comment (acikek) [inline plugins/Denizen/scripts/mobs/general_effects/shelled.dsc]: `- if not ...`
Why is this check needed, anyway?

File: plugins/Denizen/scripts/mobs/general_effects/shelled.dsc [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Shelled enemies will absorb a number of hits (specifically hits, not damage) and then break.
```
Comment (acikek) [inline plugins/Denizen/scripts/mobs/general_effects/shelled.dsc]: And then their shells will break?

File: plugins/Denizen/scripts/mobs/modifiers/assignment.dsc [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+            - foreach <[selected_prefixes]> as:prefix:
+                - flag <context.entity> mob_modifiers.<[prefix]>
+            - foreach <[selected_suffixes]> as:suffix:
+                - flag <context.entity> mob_modifiers.<[suffix]>
+            # Set the entity's custom name.
+            - define color <&color[<[build].get[color]>]>
+            - define formatted_prefixes <[selected_prefixes].parse_tag[<[parse_value].proc[snake_case_to_spaced_title_case]>].space_separated>
+            - define formatted_suffixes <[selected_suffixes].parse_tag[<[parse_value].proc[snake_case_to_spaced_title_case]>].space_separated>
+            - define entity_name <context.entity.custom_name.if_null[<context.entity.entity_type.proc[snake_case_to_spaced_title_case]>]>
+            - define name <[color]><[formatted_prefixes]> <&f><[entity_name]> The <[color]><[formatted_suffixes]>
```
Comment (acikek) [inline plugins/Denizen/scripts/mobs/modifiers/assignment.dsc]: Maybe lowercase `the`

File: plugins/Denizen/scripts/mobs/modifiers/chancing.dsc [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    debug: false
+    description: Gets a list of valid suffixes for a given build.
+    definitions: build[build to match against]
+    script:
+        - define suffixes <server.flag[mob_modifiers.suffixes]>
+        - define valid_suffixes <list>
+        - foreach <[suffixes]> as:suffix:
+            - if not <[builds].get[blacklisted_modifiers].contains[<[suffix]>]>:
+                - define valid_suffixes:->:<[suffix]>
+        - determine <[valid_suffixes]>
```
Comment (acikek) [inline plugins/Denizen/scripts/mobs/modifiers/chancing.dsc]: Essentially copypaste, maybe create another task to abstract the logic?

File: plugins/Denizen/scripts/mobs/modifiers/assignment.dsc (hunk 6 above)
Comment (funkychicken493) [reply plugins/Denizen/scripts/mobs/modifiers/assignment.dsc]: Rest of the name is uppercase.

File: plugins/Denizen/scripts/mobs/general_effects/shelled.dsc (hunk 4 above)
Comment (funkychicken493) [reply plugins/Denizen/scripts/mobs/general_effects/shelled.dsc]: Bad habit of random data checking, will fix.

File: plugins/Denizen/scripts/mobs/modifiers/chancing.dsc (hunk 7 above)
Comment (funkychicken493) [reply plugins/Denizen/scripts/mobs/modifiers/chancing.dsc]: I blame Copilot. Will fix.

File: plugins/Denizen/scripts/mobs/modifiers/assignment.dsc (hunk 6 above)
Comment (acikek) [reply plugins/Denizen/scripts/mobs/modifiers/assignment.dsc]: English title case demands `Word the Word`

**Review by acikek — APPROVED** [review]

File: plugins/Denizen/scripts/mobs/modifiers/assignment.dsc (hunk 6 above)
Comment (funkychicken493) [reply plugins/Denizen/scripts/mobs/modifiers/assignment.dsc]: Screw English, there are no rules!

**Comment by funkychicken493** [comment]
Currently in testing phase, will give update once completely debugged.

**Comment by funkychicken493** [comment]
Debugged, working, and ready for pull @acikek

**Pull request merged**